File: 3sumclosest/solve.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        sort(nums.begin(), nums.end());

        int res = 1000000;

        for(int i = 0; i < n; i++){
            if(i > 0 && nums[i] == nums[i - 1]) continue;
            int j = i + 1, k = n - 1;
            while(j < k){
                int total = nums[i] + nums[j] + nums[k];
                
                if(res == 1000000 || (abs(target-total) < abs(target-res)))
                    res = total;
                
                if(total > target) k--;
                else if(total < target) j++;
                else return total;
                
            }
        }

        return res;
    }
};

int main(){
    
    vector<int> nums = {-1, 2, 1, -4};
    int target = 1;

    Solution s;
    cout << s.threeSumClosest(nums, target);

    return 0;
}

File: 3sumclosest/solve.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        n = len(nums)
        nums.sort()

        res = 1000000

        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j, k = i + 1, n - 1
            while j < k:
                total = nums[i] + nums[j] + nums[k]
                if res == 1000000 or abs(target - total) < abs(target - res):
                    res = total
                if total > target: k -= 1
                elif total < target: j += 1
                else: return target
        return res
    
if __name__ == '__main__':
    nums = [-1, 2, 1, -4]
    target = 1
    print(Solution().threeSumClosest(nums, target))
        